velox/exec/GroupingSet.cpp

```diff
@@ -412,4 +444,241 @@ const HashLookup& GroupingSet::hashLookup() const {
+      outOfLineBytes - (rowsToSpill * outOfLineBytesPerRow));
+}
+
+void GroupingSet::spill(int64_t targetRows, int64_t targetBytes) {
+  if (!spiller_) {
+    auto rows = table_->rows();
+    auto types = rows->keyTypes();
+    types.insert(
+        types.end(), intermediateTypes_.begin(), intermediateTypes_.end());
+    std::vector<std::string> names;
+    for (auto i = 0; i < types.size(); ++i) {
+      names.push_back(fmt::format("s{}", i));
+    }
+    auto fileSize = mappedMemory_->tracker()->getCurrentUserBytes() / 4;
+    spiller_ = std::make_unique<Spiller>(
+        *rows,
+        [&](folly::Range<char**> rows) { table_->erase(rows); },
+        ROW(std::move(names), std::move(types)),
+        // Spill up to 4 partitions based on bits 29 and 30 of the hash number.
+        HashBitRange(29, 31),
+        rows->keyTypes().size(),
+        spillPath_.value(),
+        fileSize,
+        Spiller::spillPool(),
+        spillExecutor_);
+  }
+  spiller_->spill(targetRows, targetBytes, spillIterator_);
+}
+
+bool GroupingSet::getOutputWithSpill(const RowVectorPtr& result) {
+  if (outputPartition_ == -1) {
+    mergeArgs_.resize(1);
+    std::vector<TypePtr> keyTypes;
+    for (auto& hasher : table_->hashers()) {
+      keyTypes.push_back(hasher->type());
+    }
+    mergeRows_ = std::make_unique<RowContainer>(
+        keyTypes,
+        !ignoreNullKeys_,
+        aggregates_,
+        std::vector<TypePtr>(),
+        false,
+        false,
+        false,
+        false,
+        mappedMemory_,
+        ContainerRowSerde::instance());
+    // Take ownership of the rows and free the hash table. The table will not be
+    // needed for producing spill output.
+    rowsWhileReadingSpill_ = table_->moveRows();
+    table_.reset();
+    outputPartition_ = 0;
+    nonSpilledRows_ = spiller_->finishSpill();
+  }
+
+  if (nonSpilledIndex_ < nonSpilledRows_.value().size()) {
+    uint64_t bytes = 0;
+    vector_size_t numGroups = 0;
+    auto limit = std::min<size_t>(
+        1000, nonSpilledRows_.value().size() - nonSpilledIndex_);
+    for (; numGroups < limit; ++numGroups) {
+      bytes += table_->rows()->rowSize(
+          nonSpilledRows_.value()[nonSpilledIndex_ + numGroups]);
+      if (bytes > maxBatchBytes_) {
+        ++numGroups;
+        break;
+      }
+    }
+    extractGroups(
+        folly::Range<char**>(
+            nonSpilledRows_.value().data() + nonSpilledIndex_, numGroups),
+        result);
+    nonSpilledIndex_ += numGroups;
+    return true;
+  }
+  while (outputPartition_ < spiller_->state().numPartitions()) {
+    if (!merge_) {
+      merge_ = spiller_->startMerge(outputPartition_);
+    }
+    if (!mergeNext(result)) {
+      ++outputPartition_;
+      merge_ = nullptr;
+      continue;
+    }
+    return true;
+  }
+  return false;
+}
+
+bool GroupingSet::mergeNext(const RowVectorPtr& result) {
+  constexpr int32_t kBatchBytes = 1 << 20; // 1MB
+  if (mergeSelection_.size() < 64) {
```

Comment:
Why 1MB and 64? Please, comment to explain. It would be nice to collect all the constants to a single place and document these together.